docs/SPIR-V.rst

```diff
@@ -0,0 +1,194 @@
+                     V
+                SPIRVEmitter
+                     |
+                     | contains
+                     |
+       +-------------+------------+
+       |                          |
+       |                          |
+       V         references       V
+  SPIRVContext <------------ ModuleBuilder
+                                  |
+                                  | contains
+                                  V
+                              InstBuilder
+                                  |
+                                  | depends on
+                                  V
+                             WordConsumer
+
+``SPIRVEmitter``
+  The derived ``ASTConsumer`` which acts on various frontend AST nodes by calling corresponding ``ModuleBuilder`` methods to build SPIR-V modules gradually.
+``ModuleBuilder``
+  Exposes API for constructing SPIR-V modules. Interally it has structured representation of SPIR-V modules, functions, basic blocks as well as various SPIR-V specific structs like entry points, debug names, and so on.
+``SPIRVContext``
+  Responsible for <result-id> allocation and maintaining the lifetime of objects allocated to represent types, decorations, and others. It is used in conjunction with ``ModuleBuilder``.
+``InstBuilder``
+  The low-level interface for generating SPIR-V words for various SPIR-V instructions. All SPIR-V instructions are eventually serialized via ``InstBuilder``.
+``WordConsumer``
+  The consumer of generated SPIR-V words.
+
+Command-line tool
+-----------------
+
+On the command-line tool side, this means introducing a new binary, ``hlsl2spirv`` to wrap around the library functionality.
+
+But as the initial scaffolding step, a new option, ``-spirv``, will be added into ``dxc`` for invoking the new SPIR-V codegen action.
+
+Build system
+------------
+
+SPIR-V codegen functionality will structured as an optional feature in DirectXShaderCompiler. Two new CMake options will be introduced to control the configuring and building SPIR-V codegen:
+
+``ENABLE_SPIRV_CODEGEN``
+  If turned on, enables the SPIR-V codegen functionality.
+``SPIRV_BUILD_TESTS``
+  If turned on, enables building of SPIR-V related tests. This option will also implicitly turns on ``ENABLE_SPIRV_CODEGEN``.
+
+For the ``hctbuild`` script, the corresponding switches for the above two options are ``-spirv`` and ``-spirvtest``.
+
+LLVM optimization passes
+------------------------
+
+Translating frontend AST directly into SPIR-V binary precludes the usage of existing LLVM optimization passes. This is expected since there are also vast semantics differences between SPIR-V and LLVM IR. Certain concepts in SPIR-V do not have direct corresponding representation in LLVM IR and there are no existing translation schemes handling the differences. Using vanilla LLVM optimization passes will likely violate the requirements of SPIR-V and results in invalid SPIR-V modules.
+
+Mapping between HLSL and SPIR-V
```

Comment:
I expect this whole section will become reference material for people writing HLSL code.  They will not care about how you designed the compiler internals.  Eventually this reference section will become its own separate document.  Does not have to be now.